src/ModelContextProtocol.AspNetCore/Authentication/McpAuthenticationHandler.cs

```diff
@@ -0,0 +1,149 @@
+        string requestPath = Request.Path.Value ?? string.Empty;
+        var options = _optionsMonitor.CurrentValue;
+        string resourceMetadataPath = options.ResourceMetadataUri.ToString();
+        
+        // If the path doesn't match, let the request continue through the pipeline
+        if (!string.Equals(requestPath, resourceMetadataPath, StringComparison.OrdinalIgnoreCase))
+        {
+            return false;
+        }
+
+        // This is a request for resource metadata - handle it
+        await HandleResourceMetadataRequestAsync();
+        return true;
+    }
+
+    /// <summary>
+    /// Gets the base URL from the current request, including scheme, host, and path base.
+    /// </summary>
+    private string GetBaseUrl() => $"{Request.Scheme}://{Request.Host}{Request.PathBase}";
+
+    /// <summary>
+    /// Gets the absolute URI for the resource metadata endpoint.
+    /// </summary>
+    private string GetAbsoluteResourceMetadataUri()
+    {
+        var options = _optionsMonitor.CurrentValue;
+        var resourceMetadataUri = options.ResourceMetadataUri;
+        
+        if (resourceMetadataUri.IsAbsoluteUri)
+        {
+            return resourceMetadataUri.ToString();
+        }
+        
+        // For relative URIs, combine with the base URL
+        string baseUrl = GetBaseUrl();
+        string resourceMetadataPath = resourceMetadataUri.ToString();
+        
+        if (!Uri.TryCreate(baseUrl + resourceMetadataPath, UriKind.Absolute, out var absoluteUri))
+        {
+            throw new InvalidOperationException("Could not create absolute URI for resource metadata.");
+        }
+        
+        return absoluteUri.ToString();
+    }
+
+    /// <summary>
+    /// Handles the resource metadata request.
+    /// </summary>
+    private async Task HandleResourceMetadataRequestAsync()
+    {
+        // Get resource metadata from options, using the dynamic provider if available
+        var options = _optionsMonitor.CurrentValue;
+        var resourceMetadata = options.GetResourceMetadata(Request.HttpContext);
+        
+        // Create a copy to avoid modifying the original
+        var metadata = new ProtectedResourceMetadata
+        {
+            AuthorizationServers = [.. resourceMetadata.AuthorizationServers],
+            BearerMethodsSupported = [.. resourceMetadata.BearerMethodsSupported],
+            ScopesSupported = [.. resourceMetadata.ScopesSupported],
+            ResourceDocumentation = resourceMetadata.ResourceDocumentation,
+            Resource = resourceMetadata.Resource
+        };
+        
+        // Set default resource if not set
+        if (metadata.Resource == null)
+        {
+            metadata.Resource = new Uri(GetBaseUrl());
+        }
+        
+        Response.StatusCode = StatusCodes.Status200OK;
+        Response.ContentType = "application/json";
+        
+        var json = System.Text.Json.JsonSerializer.Serialize(
```

Comment:
Can we have a `using System.Text.Json;` and then remove the namespace from here?